### PR TITLE
Object detection experiment for transformers

### PR DESCRIPTION
**Pull request by SangbumChoi**
# What does this PR do?

@qubvel As we discussed, I will generate the standard metric for 

**how well the transformers object detection model finetune in diverse domain of data?**

In this step I will focus on 

batch_size | learning rate | model | dataset | AP | training time | required memory
-- | -- | -- | -- | -- | -- | --

on different domain of dataset.
https://www.notion.so/sangbumchoi/HuggingFace-OD-train-1afb8ded2e1680daa2dcfdc3e11aa15d?pvs=4

For the part that changed is

In this step D-Fine was not perfectly finished for the finetune feature. You can see the https://github.com/Peterande/D-FINE/blob/4a1f73a8bcfac736a88abde9596d87f116d780a7/src/zoo/dfine/dfine_criterion.py#L357. This is `pre_output` related auxiliary layer which is related to first layer to use `pre_bbox_head` .https://github.com/huggingface/transformers/blob/2932f318a20d9e54cc7aea052e040164d85de7d6/src/transformers/models/d_fine/modeling_d_fine.py#L1195

This is initial experiment of four models that you have suggested.

Screenshot 2025-05-05 at 2 54 56 PM

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by qubvel — COMMENTED** [review]
Hey @SangbumChoi, thanks for opening a PR! Can you please make a separate PR for D-Fine fixes?

**Comment by qubvel** [comment]
cc @VladOS95-cyber

File: src/transformers/models/d_fine/modeling_d_fine.py [hunk 1]
```diff
@@ -1293,8 +1311,9 @@ def forward(
 
             if i == 0:
                 # Initial bounding box predictions with inverse sigmoid refinement
-                new_reference_points = F.sigmoid(self.pre_bbox_head(output[0]) + inverse_sigmoid(ref_points_detach))
-                ref_points_initial = new_reference_points.detach()
+                initial_bboxes = F.sigmoid(self.pre_bbox_head(hidden_states) + inverse_sigmoid(ref_points_detach))
+                initial_scores = self.class_embed[0](hidden_states)
+                ref_points_initial = initial_bboxes.detach()
```
[the remark this answers is not in the document]
Comment (SangbumChoi) [reply src/transformers/models/d_fine/modeling_d_fine.py:1316]: 1. for the name part you are correct there is no need to change it
2. however since the first part of ref_points_initial is detached so it does not flow the gradient to the model architecture, but the original code was intend to use the first layer output named as `pre-output` and use this in to a auxiliary loss function

As a result I can change the name to original `new_reference_points` but in order to use the initial. How about change `initial_scores` to `new_reference_scores`?

@qubvel

**Comment by SangbumChoi** [comment]
For the PR I will break into two parts :)

File: src/transformers/models/d_fine/modeling_d_fine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/d_fine/modeling_d_fine.py:1316]: I would say it's ok to change naming and correct logic if there is a bug, but I would avoid adding new model outputs if we can avoid it. In addition, please make sure its aligned with RT-DETR model, cause RT-DETR is used as a modular base

[the remark this answers is not in the document]
Comment (SangbumChoi) [reply src/transformers/models/d_fine/modeling_d_fine.py:1316]: Yeah I will make the another PR with the modified modular_d_fine.py
